### ISSUE TITLE
redo apply failed due to "panic: insert events should be emitted after other kinds of events in the same transaction"

### ISSUE THREAD

**Issue by fubinzh** [pingcap/tiflow#11321]
### What did you do?

1. create mysql sink changefeed with eventual consistency on.
2. run workload
```
/go-tpc tpcc --db workload --warehouses 200 -T 50 --host upstream-tidb.cdc-redo-long-duration-tps-7589946-1-17 --port 4000 --parts 1 prepare --ignore-error '2013,1213,1105,1205,8022,8028,9004,9007,1062'
/go-tpc tpcc --db workload --warehouses 200 -T 50 --host upstream-tidb.cdc-redo-long-duration-tps-7589946-1-17 --port 4000 --parts 1 --time 24h0m0s run --ignore-error '2013,1213,1105,1205,8022,8028,9004,9007,1062'
```
3. after run workload for 30m, pause changefeed and run run redo apply 
```
cdc  redo  apply "--sink-uri=mysql://root:@downstream.cdc-redo-long-duration-tps-7589946-1-17:3306" "--storage=s3://tmp/test-infra-redolog/redo-apply-multiple-cplcs3i4eoooehph03j0-2024-06-13 10:46:38.371569253 +0000 UTC m=+31.379610414?access-key=xxx&secret-access-key=xxx&endpoint=http://minio-peer:9000&force-path-style=true" "--tmp-dir=/tmp/nfs/redo-apply-multiple-cplcs3i4eoooehph03j02024-06-13-10-46-38"
```

### What did you expect to see?

redo apply should succeed.

### What did you see instead?

redo apply failed.

```
/cdc  redo  apply \"--sink-uri=mysql://root:@downstream.cdc-redo-long-duration-tps-7589946-1-17:3306\" \"--storage=s3://tmp/test-infra-redolog/redo-apply-multiple-cplcs3i4eoooehph03j0-2024-06-13 10:46:38.371569253 +0000 UTC m=+31.379610414?access-key=xxx&secret-access-key=xxx&endpoint=http://minio-peer:9000&force-path-style=true\" \"--tmp-dir=/tmp/nfs/redo-apply-multiple-cplcs3i4eoooehph03j02024-06-13-10-46-38\", stdout: [2024/06/13 11:32:39.927 +00:00] [INFO] [helper.go:50] [\"init log\"] [file=] [level=info]\n[2024/06/13 11:32:39.929 +00:00] [INFO] [apply.go:78] [\"set memory limit\"] [memoryLimit=10737418240]\n[2024/06/13 11:32:39.929 +00:00] [INFO] [apply.go:94] [\"Start http pprof server\"] [addr=:6060]\n[2024/06/13 11:32:39.933 +00:00] [INFO] [s3.go:399] [\"succeed to get bucket region from s3\"] [\"bucket region\"=us-east-1]\n[2024/06/13 11:32:39.945 +00:00] [INFO] [mem_quota.go:80] [\"New memory quota\"] [namespace=default] [changefeed=redo-applier] [total=1073741824]\n[2024/06/13 11:32:39.945 +00:00] [INFO] [redo.go:163] [\"apply redo log starts\"] [checkpointTs=450436360840675585] [resolvedTs=450436361561571490]\n[2024/06/13 11:32:39.947 +00:00] [INFO] [sink.go:417] [\"succeed to parse parameter from sink uri\"] [protocol=] [txnAtomicity=]\n[2024/06/13 11:32:39.947 +00:00] [WARN] [config.go:374] [\"Because time-zone is not specified, the timezone of the TiCDC server will be used. We recommend that you specify the time-zone explicitly. Please make sure that the timezone of the TiCDC server, sink-uri and the downstream database are consistent. If the downstream database does not load the timezone information, you can refer to https://dev.mysql.com/doc/refman/8.0/en/mysql-tzinfo-to-sql.html.\"] [timezone=UTC]\n[2024/06/13 11:32:39.948 +00:00] [INFO] [s3.go:399] [\"succeed to get bucket region from s3\"] [\"bucket region\"=us-east-1]\n[2024/06/13 11:32:39.949 +00:00] [INFO] [s3.go:399] [\"succeed to get bucket region from s3\"] [\"bucket region\"=us-east-1]\n[2024/06/13 11:32:39.957 +00:00] [INFO] [file.go:113] [\"succeed to download and sort redo logs\"] [type=ddl] [duration=11.268234ms]\n[2024/06/13 11:32:39.958 +00:00] [INFO] [db_helper.go:175] [\"sink uri is configured\"] [dsn=\"root:******@tcp(downstream.cdc-redo-long-duration-tps-7589946-1-17:3306)/?interpolateParams=true&multiStatements=true&charset=utf8mb4&foreign_key_checks=0&maxAllowedPacket=0&readTimeout=2m&sql_mode=%22NO_ENGINE_SUBSTITUTION%2CIGNORE_SPACE%2CNO_AUTO_VALUE_ON_ZERO%2CONLY_FULL_GROUP_BY%2CALLOW_INVALID_DATES%22&time_zone=%22UTC%22&timeout=2m&transaction_isolation=%22READ-COMMITTED%22&writeTimeout=2m\"]\n[2024/06/13 11:32:39.962 +00:00] [ERROR] [db_helper.go:320] [\"check tidb version error\"] [error=\"Error 1046 (3D000): No database selected\"]\n[2024/06/13 11:32:39.962 +00:00] [ERROR] [db_helper.go:320] [\"check tidb version error\"] [error=\"Error 1046 (3D000): No database selected\"]\n[2024/06/13 11:32:39.963 +00:00] [INFO] [mysql.go:141] [\"MySQL backends is created\"] [changefeed=.] [workerCount=16] [forceReplicate=false] [enableOldValue=true]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.966 +00:00] [INFO] [txn_cache.go:64] [\"create new worker cache in conflict detector\"] [cacheCount=16] [cacheSize=1024] [BlockStrategy=waitAll]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=0]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=2]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=5]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=3]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=1]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=4]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=7]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=6]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=8]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=9]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=10]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=11]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=12]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=13]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [sink.go:417] [\"succeed to parse parameter from sink uri\"] [protocol=] [txnAtomicity=]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=14]\n[2024/06/13 11:32:39.967 +00:00] [INFO] [worker.go:89] [\"Transaction dmlSink worker starts\"] [changefeedID=.] [workerID=15]\n[2024/06/13 11:32:39.967 +00:00] [WARN] [config.go:374] [\"Because time-zone is not specified, the timezone of the TiCDC server will be used. We recommend that you specify the time-zone explicitly. Please make sure that the timezone of the TiCDC server, sink-uri and the downstream database are consistent. If the downstream database does not load the timezone information, you can refer to https://dev.mysql.com/doc/refman/8.0/en/mysql-tzinfo-to-sql.html.\"] [timezone=UTC]\n[2024/06/13 11:32:39.977 +00:00] [INFO] [db_helper.go:175] [\"sink uri is configured\"] [dsn=\"root:******@tcp(downstream.cdc-redo-long-duration-tps-7589946-1-17:3306)/?interpolateParams=true&multiStatements=true&charset=utf8mb4&foreign_key_checks=0&maxAllowedPacket=0&readTimeout=2m&sql_mode=%22IGNORE_SPACE%2CNO_AUTO_VALUE_ON_ZERO%2CONLY_FULL_GROUP_BY%2CALLOW_INVALID_DATES%2CNO_ENGINE_SUBSTITUTION%22&time_zone=%22UTC%22&timeout=2m&transaction_isolation=%22READ-COMMITTED%22&writeTimeout=2m\"]\n[2024/06/13 11:32:39.983 +00:00] [ERROR] [db_helper.go:320] [\"check tidb version error\"] [error=\"Error 1046 (3D000): No database selected\"]\n[2024/06/13 11:32:39.983 +00:00] [INFO] [mysql_ddl_sink.go:101] [\"MySQL DDL sink is created\"] [namespace=] [changefeed=]\n[2024/06/13 11:32:41.248 +00:00] [INFO] [file.go:283] [\"ignore logs which commitTs is greater than resolvedTs\"] [filename=upstream-ticdc-1.upstream-ticdc-peer.cdc-redo-long-duration-tps-7589946-1-17.svc:8301_redo-apply-multiple-cplcs3i4eoooehph03j0_row_450436361640214531_b5c36212-59c2-4acd-a88a-cd40baaa708a.log] [endTs=450436361561571490]\n[2024/06/13 11:32:41.331 +00:00] [INFO] [file.go:283] [\"ignore logs which commitTs is greater than resolvedTs\"] [filename=upstream-ticdc-0.upstream-ticdc-peer.cdc-redo-long-duration-tps-7589946-1-17.svc:8301_redo-apply-multiple-cplcs3i4eoooehph03j0_row_450436361627369655_b9f6dd23-2666-4e99-93ac-fd48ca21acbd.log] [endTs=450436361561571490]\n[2024/06/13 11:32:41.361 +00:00] [INFO] [file.go:113] [\"succeed to download and sort redo logs\"] [type=row] [duration=1.415545396s]\n[2024/06/13 11:32:41.361 +00:00] [PANIC] [redo.go:653] [\"insert events should be emitted after other kinds of events in the same transaction\"] [stack=\"github.com/pingcap/tiflow/pkg/applier.(*updateEventSplitter).checkEventOrder\\n\\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:653\\ngithub.com/pingcap/tiflow/pkg/applier.(*updateEventSplitter).readNextRow\\n\\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:693\\ngithub.com/pingcap/tiflow/pkg/applier.(*RedoApplier).consumeLogs\\n\\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:206\\ngithub.com/pingcap/tiflow/pkg/applier.(*RedoApplier).Apply.func3\\n\\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:741\\ngolang.org/x/sync/errgroup.(*Group).Go.func1\\n\\tgolang.org/x/sync@v0.7.0/errgroup/errgroup.go:78\"]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=0]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=6]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=11]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=1]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=9]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=7]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=2]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=3]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=8]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=4]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=10]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=13]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=5]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=15]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=14]\n[2024/06/13 11:32:41.362 +00:00] [INFO] [worker.go:100] [\"Transaction dmlSink worker exits as canceled\"] [changefeedID=.] [workerID=12], stderr: panic: insert events should be emitted after other kinds of events in the same transaction\n\ngoroutine 465 [running]:\ngo.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x0?, 0x0?, {0x0?, 0x0?, 0xc000512740?})\n\tgo.uber.org/zap@v1.26.0/zapcore/entry.go:196 +0x65\ngo.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000e5a340, {0x0, 0x0, 0x0})\n\tgo.uber.org/zap@v1.26.0/zapcore/entry.go:262 +0x3ec\ngo.uber.org/zap.(*Logger).Panic(0xc000e8f770?, {0x3fb7159?, 0x33d2aa5?}, {0x0, 0x0, 0x0})\n\tgo.uber.org/zap@v1.26.0/logger.go:284 +0x59\ngithub.com/pingcap/log.Panic({0x3fb7159?, 0xc000e53500?}, {0x0?, 0xc00067a2a0?, 0x0?})\n\tgithub.com/pingcap/log@v1.1.1-0.20230317032135-a0d097d16e22/global.go:54 +0x8b\ngithub.com/pingcap/tiflow/pkg/applier.(*updateEventSplitter).checkEventOrder(0xc000ebf3a0?, 0x48c9c10?)\n\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:653 +0x5a\ngithub.com/pingcap/tiflow/pkg/applier.(*updateEventSplitter).readNextRow(0xc000ff6b80, {0x48c9c10, 0xc0005cd8c0})\n\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:693 +0xb7\ngithub.com/pingcap/tiflow/pkg/applier.(*RedoApplier).consumeLogs(0xc00066fd50, {0x48c9c10, 0xc0005cd8c0})\n\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:206 +0x456\ngithub.com/pingcap/tiflow/pkg/applier.(*RedoApplier).Apply.func3()\n\tgithub.com/pingcap/tiflow/pkg/applier/redo.go:741 +0x25\ngolang.org/x/sync/errgroup.(*Group).Go.func1()\n\tgolang.org/x/sync@v0.7.0/errgroup/errgroup.go:78 +0x64\ncreated by golang.org/x/sync/errgroup.(*Group).Go\n\tgolang.org/x/sync@v0.7.0/errgroup/errgroup.go:75 +0xa5, ExitCode: 2"] [stack="github.com/pingcap/test-infra/caselib/pkg/host.(*TiCDCHost).Redo\n\tgithub.com/pingcap/test-infra/caselib/pkg/host/ticdc.go:107\ngithub.com/pingcap/test-infra/caselib/pkg/steps.(*redo).Execute\n\tgithub.com/pingcap/test-infra/caselib/pkg/steps/redo.go:28\ngithub.com/pingcap/test-infra/caselib/pkg/steps.withRecover\n\tgithub.com/pingcap/test-infra/caselib/pkg/steps/step.go:24\ngithub.com/pingcap/test-infra/caselib/pkg/steps.(*Serial).Execute\n\tgithub.com/pingcap/test-infra/caselib/pkg/steps/step.go:45\nmain.main\n\ttest-infra/caselib/ticdc/main.go:112\nruntime.main\n\truntime/proc.go:267"]
[2024/06/13 11:32:41.479 +00:00] [INFO] [redo.go:29] ["run redo cmd finished"] [result=null] [error="command terminated with exit code 2"] [errorVerbose="command terminated with exit code 2\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/test-infra/sdk/resource/impl/k8s.(*TiDBCluster).Exec\n\tgithub.com/pingcap/test-infra/sdk@v0.0.0-20240306052032-3cff273d546b/resource/impl/k8s/tidbcluster.go:199\ngithub.com/pingcap/test-infra/caselib/pkg/host.(*TiCDCHost).Exec.func1\n\tgithub.com/pingcap/test-infra/caselib/pkg/host/ticdc.go:303\nruntime.goexit\n\truntime/asm_amd64.s:1650"]
[2024/06/13 11:32:41.479 +00:00] [PANIC] [main.go:113] ["run case failed"] [error="command terminated with exit code 2"] [errorVerbose="command terminated with exit code 2\ngithub.com/pingcap/errors.AddStack\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.Trace\n\tgithub.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/juju_adaptor.go:15\ngithub.com/pingcap/test-infra/sdk/resource/impl/k8s.(*TiDBCluster).Exec\n\tgithub.com/pingcap/test-infra/sdk@v0.0.0-20240306052032-3cff273d546b/resource/impl/k8s/tidbcluster.go:199\ngithub.com/pingcap/test-infra/caselib/pkg/host.(*TiCDCHost).Exec.func1\n\tgithub.com/pingcap/test-infra/caselib/pkg/host/ticdc.go:303\nruntime.goexit\n\truntime/asm_amd64.s:1650"
```

### Versions of the cluster

Git Commit Hash: 66d7a1c46e0be82311e637f4d453d3ea4dd6ea2a

**Comment by fubinzh** [comment]
This issue was supposed having been already fixed in release-7.5 branch by https://github.com/pingcap/tiflow/pull/11155, somehow it is still seen with v7.5.2 https://github.com/pingcap/tiflow/commit/66d7a1c46e0be82311e637f4d453d3ea4dd6ea2a

**Comment by hongyunyan** [comment]
duplicate with https://github.com/pingcap/tiflow/issues/11096
